// zxcvbn/lib/feedback.js
// Generated by CoffeeScript 1.10.0
var feedback, scoring;

scoring = require('./scoring');

feedback = {
  default_feedback: {
    warning: '',
    suggestions: ["Usa pocas palabras, evita frases comunes", "No hay necesidad de símbolos, dígitos o letras mayúsculas"]
  },
  get_feedback: function (score, sequence) {
    var extra_feedback, i, len, longest_match, match, ref;
    if (sequence.length === 0) {
      return this.default_feedback;
    }
    if (score > 2) {
      return {
        warning: '',
        suggestions: []
      };
    }
    longest_match = sequence[0];
    ref = sequence.slice(1);
    for (i = 0, len = ref.length; i < len; i++) {
      match = ref[i];
      if (match.token.length > longest_match.token.length) {
        longest_match = match;
      }
    }
    feedback = this.get_match_feedback(longest_match, sequence.length === 1);
    extra_feedback = 'Agregue otra palabra o dos. Las palabras no muy comunes son mejores.';
    if (feedback != null) {
      feedback.suggestions.unshift(extra_feedback);
      if (feedback.warning == null) {
        feedback.warning = '';
      }
    } else {
      feedback = {
        warning: '',
        suggestions: [extra_feedback]
      };
    }
    return feedback;
  },
  get_match_feedback: function (match, is_sole_match) {
    var layout, warning;
    switch (match.pattern) {
      case 'dictionary':
        return this.get_dictionary_match_feedback(match, is_sole_match);
      case 'spatial':
        layout = match.graph.toUpperCase();
        warning = match.turns === 1 ? 'Las filas rectas de teclas son fáciles de adivinar' : 'Los patrones cortos de teclado son fáciles de adivinar';
        return {
          warning: warning,
          suggestions: ['Usa un patrón de teclado más largo con más vueltas']
        };
      case 'repeat':
        warning = match.base_token.length === 1 ? 'Las repeticiones como "aaa" son fáciles de adivinar' : 'Las repeticiones como "abcabcabc" son solo un poco más difíciles de adivinar que "abc"';
        return {
          warning: warning,
          suggestions: ['Evite palabras y caracteres repetidos']
        };
      case 'sequence':
        return {
          warning: "Secuencias como abc o 6543 son fáciles de adivinar",
          suggestions: ['Evitar secuencias']
        };
      case 'regex':
        if (match.regex_name === 'recent_year') {
          return {
            warning: "Los años recientes son fáciles de adivinar",
            suggestions: ['Evite años recientes', 'Evite los años que están asociados con usted']
          };
        }
        break;
      case 'date':
        return {
          warning: "Las fechas son a menudo fáciles de adivinar",
          suggestions: ['Evite las fechas y años que están asociados con usted']
        };
    }
  },
  get_dictionary_match_feedback: function (match, is_sole_match) {
    var ref, result, suggestions, warning, word;
    warning = match.dictionary_name === 'passwords' ? is_sole_match && !match.l33t && !match.reversed ? match.rank <= 10 ? 'Esta es una de las 10 contraseñas comunes' : match.rank <= 100 ? 'Esta es una contraseña común de los 100 principales' : 'Esta es una contraseña muy común' : match.guesses_log10 <= 4 ? 'Esto es similar a una contraseña de uso común' : void 0 : match.dictionary_name === 'english_wikipedia' ? is_sole_match ? 'Una palabra en sí misma es fácil de adivinar' : void 0 : (ref = match.dictionary_name) === 'surnames' || ref === 'male_names' || ref === 'female_names' ? is_sole_match ? 'Los nombres y apellidos por sí mismos son fáciles de adivinar' : 'Los nombres y apellidos comunes son fáciles de adivinar' : '';
    suggestions = [];
    word = match.token;
    if (word.match(scoring.START_UPPER)) {
      suggestions.push("La mayúscula no ayuda mucho");
    } else if (word.match(scoring.ALL_UPPER) && word.toLowerCase() !== word) {
      suggestions.push("Todo en mayúsculas es casi tan fácil de adivinar como minúsculas");
    }
    if (match.reversed && match.token.length >= 4) {
      suggestions.push("Las palabras invertidas no son mucho más difíciles de adivinar");
    }
    if (match.l33t) {
      suggestions.push("Las sustituciones predecibles como '@' en lugar de 'a' no ayudan mucho");
    }
    result = {
      warning: warning,
      suggestions: suggestions
    };
    return result;
  }
};

module.exports = feedback;
